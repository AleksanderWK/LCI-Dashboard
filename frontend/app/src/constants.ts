// Frequency in Hz (cycles per second)
export const FREQUENCY = 2;

// Maximum time interval to show on live charts in seconds
export const LIVE_CHART_RANGE = 30;

export enum Variable {
    CognitiveLoad = "cl",
    PerceivedDifficulty = "pd",
    Familiarity = "fam",
    InformationProcessingIndex = "ipi",
    PhysiologicalArousal = "pa",
    Engagement = "eng",
    PhysiologicalStress = "ps",
    EmotionalRegulation = "er",
    MotionStability = "ms",
    EnergySpentFatigue = "esf",
    EducationalSpecificEmotions = "ese"
}

export enum Device {
    EyeTracker = "Eye Tracker",
    Wristband = "Wristband",
    VideoBody = "Video of Body",
    VideoFace = "Video of Face"
}

export enum EyeTrackingDevice {
    Mobile = "Mobile",
    Stationary = "Stationary"
}

export interface MMDVariableInfo {
    name: string;
    device: Device;
    description: string;
    maxValue?: number;
    calculationTime?: number;
    enabled: boolean;
}

export const MMDVariables: {[key in Variable]: MMDVariableInfo} = {
    [Variable.CognitiveLoad]: {
        name: "Cognitive Load",
        device: Device.EyeTracker,
        description:
            "Cognitive load refers to the amount of information that the working memory of a person can hold at once. In an educational setting, it is the effort needed for a learner to learn new information.",
        calculationTime: 30,
        enabled: true
    },
    [Variable.PerceivedDifficulty]: {
        name: "Perceived Difficulty",
        device: Device.EyeTracker,
        description:
            "Perceived difficulty is the perception of the ease or difficulty of performing in an educational context.",
        maxValue: 100,
        enabled: true
    },
    [Variable.Familiarity]: {
        name: "Familiarity",
        device: Device.EyeTracker,
        description: "",
        enabled: false
    },
    [Variable.InformationProcessingIndex]: {
        name: "Information Processing Index",
        device: Device.EyeTracker,
        description:
            "The ratio of local (100) and global (0) information processing. Staring at one point is local processing, while rapidly scanning a larger area is global processing.",
        maxValue: 100,
        enabled: true
    },
    [Variable.PhysiologicalArousal]: {
        name: "Physiological Arousal",
        device: Device.Wristband,
        description:
            "Physiological Arousal is how strong the current emotions are. It is calculated by the average EDA in a time interval of 8 seconds.",
        calculationTime: 8,
        enabled: true
    },
    [Variable.Engagement]: {
        name: "Engagement",
        device: Device.Wristband,
        description:
            "Engagement is a measure of involvement and enthusiasm for the activity. It is calculated by the area under the graph of tonic EDA, in a time interval of 4 seconds.",
        calculationTime: 4,
        enabled: true
    },
    [Variable.PhysiologicalStress]: {
        name: "Physiological Stress",
        device: Device.Wristband,
        description:
            "Physioloigcal Stress is indicated by LF/HF, the relative strengths of the low frequency and high frequency waves of the heart pulse signal.",
        calculationTime: 300,
        enabled: true
    },
    [Variable.EmotionalRegulation]: {
        name: "Emotional Regulation",
        device: Device.Wristband,
        description: "",
        enabled: false
    },
    [Variable.MotionStability]: {
        name: "Motion Stability",
        device: Device.VideoBody,
        description: "",
        enabled: false
    },
    [Variable.EnergySpentFatigue]: {
        name: "Energy Spent/Fatigue",
        device: Device.VideoBody,
        description:
            "The amount of energy spent by the student, also an indicator of fatigue. This is determined by calculating the jerk, the third derivative of the position skeletal body.",
        calculationTime: 3,
        enabled: true
    },
    [Variable.EducationalSpecificEmotions]: {
        name: "Educational-specific Emotions",
        device: Device.VideoFace,
        description: "Emotions that the student is experiencing in an educational setting.",
        enabled: true
    }
};

export const emotionsColorMapper: {[key: string]: string} = {
    boredom: "#F15025",
    frustration: "#191919",
    confusion: "#FFE45E",
    delight: "#6782B7"
};

export const emotionsIndexMapper: {[key: string]: number} = {
    boredom: 0,
    frustration: 1,
    confusion: 2,
    delight: 3
};

export const months = [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December"
];
