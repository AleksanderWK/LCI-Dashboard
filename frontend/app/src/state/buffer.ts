import {atom, atomFamily} from "recoil";
import {DataPoints} from "../pages/SessionView";

export const sessionCodesState = atom<string[]>({
    key: "sessionCodes",
    default: []
});

/*
 *  An atomFamily that stores buffer data for each session
 */
export const bufferState = atomFamily<{active: boolean; data: DataPoints[]}, string>({
    key: "buffer",
    default: {active: false, data: []}
});
