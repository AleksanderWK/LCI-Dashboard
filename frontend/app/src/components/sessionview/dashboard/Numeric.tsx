import React from "react";
import {createStyles, makeStyles, Theme, Typography} from "@material-ui/core";
import {useRecoilValue} from "recoil";
import {emotionsColorMapper, Variable} from "../../../constants";
import {
    selectedSessionLastValueState,
    sessionVariableDataState,
    EducationalSpecificEmotions
} from "../../../state/session";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        numericWrapper: {
            width: "100%",
            height: "100%",
            display: "flex",
            justifyContent: "center",
            alignItems: "center",
            userSelect: "none"
        },
        numeric: {
            fontSize: 70,
            color: theme.palette.text.default,
            textAlign: "center"
        },
        noData: {
            color: theme.palette.text.default,
            textAlign: "center"
        },
        emotionsContainer: {
            display: "grid",
            gridTemplateColumns: "auto",
            maxHeight: "100%",
            gap: 4
        },
        emotion: {
            display: "grid",
            gridTemplateColumns: "16px auto",
            gap: theme.spacing(1),
            alignItems: "center"
        },
        emotionIndicator: {
            width: 16,
            height: 16,
            borderRadius: 4
        },
        emotionText: {
            fontSize: 20,
            color: theme.palette.text.default
        },
        neutralText: {
            fontSize: 20,
            color: theme.palette.text.placeholder
        }
    })
);

interface EmotionsDisplayProps {
    emotions: EducationalSpecificEmotions;
}

/**
 * Displays the emotions for a specific data point.
 * @param {object} props - Component props
 * @param {EducationalSpecificEmotions} emotions - The emotions for the data point
 */
function EmotionsDisplay(props: EmotionsDisplayProps): JSX.Element {
    const classes = useStyles();

    /**
     * Maps each emotion that is true to the correct string representation
     */
    const emotions = Object.entries(props.emotions)
        .filter(([, value]) => value)
        .map(([emotion]) => {
            let text = null;

            switch (emotion) {
                case "boredom":
                    text = "Bored";
                    break;
                case "frustration":
                    text = "Frustrated";
                    break;
                case "confusion":
                    text = "Confused";
                    break;
                default:
                    text = "Delighted";
            }

            return (
                <div key={emotion} className={classes.emotion}>
                    <div
                        className={classes.emotionIndicator}
                        style={{backgroundColor: emotionsColorMapper[emotion]}}
                    ></div>
                    <Typography className={classes.emotionText}>{text}</Typography>
                </div>
            );
        });

    return (
        <div className={classes.emotionsContainer}>
            {emotions.length > 0 ? emotions : <Typography className={classes.neutralText}>Neutral</Typography>}
        </div>
    );
}

interface NumericProps {
    variable: Variable;
    id?: number;
}

/**
 * Displays the last value received for a variable.
 * @param {object} props - Component props
 * @param {Variable} props.variable - The variable to get the last value for
 * @param {number} props.id - The session ID of the session this numeric applies to.
 * If not set, the selected session is used.
 */
function Numeric(props: NumericProps): JSX.Element {
    const classes = useStyles();

    const dataPoint = !props.id
        ? useRecoilValue(selectedSessionLastValueState(props.variable))
        : useRecoilValue(sessionVariableDataState([props.variable, props.id])).slice(-1)[0][1];

    // If the variable is Educational-specific Emotions, EmotionsDisplay is used, else, a number is rendered
    return (
        <div className={`${"noDrag"} ${classes.numericWrapper}`}>
            {dataPoint && props.variable !== Variable.EducationalSpecificEmotions ? (
                <Typography className={classes.numeric}>
                    {dataPoint < 1 && dataPoint > 0
                        ? (dataPoint as number).toFixed(1)
                        : Math.round(dataPoint as number)}
                </Typography>
            ) : dataPoint && props.variable == Variable.EducationalSpecificEmotions ? (
                <EmotionsDisplay emotions={dataPoint as EducationalSpecificEmotions} />
            ) : (
                <Typography className={classes.noData}>No data received</Typography>
            )}
        </div>
    );
}

export default Numeric;
