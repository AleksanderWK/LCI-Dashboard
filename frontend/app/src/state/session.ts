import {Layouts} from "react-grid-layout";
import {atom, atomFamily, selector, selectorFamily, useRecoilValue} from "recoil";
import {EyeTrackingDevice, Variable} from "../constants";
import {studentState, Student} from "./student";

/*
 *  An atom that stores which session is selected
 */
export const selectedSessionIdState = atom<number | null>({
    key: "selectedSessionId",
    default: null
});

/*
 *  An atom that stores ids for all sessions
 */
export const sessionIdsState = atom<number[]>({
    key: "sessionIds",
    default: []
});

export interface Session {
    _id: number;
    sessionName: string;
    studentId: string;
    eyeTrackingDevice: EyeTrackingDevice;
    startTime: number;
    endTime: number | null;
    sessionCode: string;
}

/*
 *  An atomFamily that stores session information for each session
 */
export const sessionState = atomFamily<Session | undefined, number | null>({
    key: "session",
    default: undefined
});

interface Recording {
    status: boolean;
    startTime: Date | null;
    recordingId: number | null;
}

/*
 *   An atomFamily that stores the recording status of each session
 */
export const sessionRecordingState = atomFamily<Recording, number | null>({
    key: "sessionRecording",
    default: {
        status: false,
        startTime: null,
        recordingId: null
    }
});

/*
 *   A selector that returns whether the selected session is being recorded
 */
export const selectedSessionRecordingState = selector<Recording>({
    key: "selectedSessionRecording",
    get: ({get}) => {
        const id = get(selectedSessionIdState);

        return get(sessionRecordingState(id));
    },
    set: ({get, set}, newValue) => {
        const id = get(selectedSessionIdState);
        set(sessionRecordingState(id), newValue);
    }
});

export interface SessionWithStudent {
    _id: number;
    sessionName: string;
    student: Student;
    eyeTrackingDevice: EyeTrackingDevice;
    recording: Recording;
    startTime: number;
    endTime: number | null;
    sessionCode: string;
}

/*
 *  A selector that returns the session info of the selected session
 */
export const selectedSessionState = selector<SessionWithStudent | undefined>({
    key: "selectedSession",
    get: ({get}) => {
        const id = get(selectedSessionIdState);
        const session = get(sessionState(id));

        if (id && session) {
            const student = get(studentState(session.studentId));

            if (student) {
                return {
                    _id: id,
                    sessionName: session.sessionName,
                    student: student,
                    eyeTrackingDevice: session.eyeTrackingDevice,
                    recording: get(sessionRecordingState(id)),
                    startTime: session.startTime,
                    endTime: session.endTime,
                    sessionCode: session.sessionCode
                };
            }
        }
    }
});

/*
 *  A selector that returns the session info of all sessions
 */
export const sessionsState = selector<SessionWithStudent[]>({
    key: "sessions",
    get: ({get}) => {
        const sessions: SessionWithStudent[] = [];
        const sessionIds = get(sessionIdsState);

        sessionIds.forEach((sessionId) => {
            const session = get(sessionState(sessionId));
            if (session) {
                const student = get(studentState(session.studentId));

                if (student) {
                    sessions.push({
                        _id: sessionId,
                        sessionName: session.sessionName,
                        student: student,
                        eyeTrackingDevice: session.eyeTrackingDevice,
                        recording: get(sessionRecordingState(sessionId)),
                        startTime: session.startTime,
                        endTime: session.endTime,
                        sessionCode: session.sessionCode
                    });
                }
            }
        });

        return sessions;
    }
});

export interface Data {
    [Variable.CognitiveLoad]: [number, number][];
    [Variable.PerceivedDifficulty]: [number, number][];
    [Variable.Familiarity]: [number, number][];
    [Variable.InformationProcessingIndex]: [number, number][];
    [Variable.PhysiologicalArousal]: [number, number][];
    [Variable.Engagement]: [number, number][];
    [Variable.PhysiologicalStress]: [number, number][];
    [Variable.EmotionalRegulation]: [number, number][];
    [Variable.MotionStability]: [number, number][];
    [Variable.EnergySpentFatigue]: [number, number][];
}

/*
 *   An atomFamily that stores the data for each session
 */
export const sessionDataState = atomFamily<Data, number | null>({
    key: "sessionData",
    default: {
        [Variable.CognitiveLoad]: [],
        [Variable.PerceivedDifficulty]: [],
        [Variable.Familiarity]: [],
        [Variable.InformationProcessingIndex]: [],
        [Variable.PhysiologicalArousal]: [],
        [Variable.Engagement]: [],
        [Variable.PhysiologicalStress]: [],
        [Variable.EmotionalRegulation]: [],
        [Variable.MotionStability]: [],
        [Variable.EnergySpentFatigue]: []
    }
});

/*
 *  A selector that returns the selected session's data
 */
export const selectedSessionDataState = selector<Data>({
    key: "selectedSessionData",
    get: ({get}) => {
        const id = get(selectedSessionIdState);

        return get(sessionDataState(id));
    },
    set: ({get, set}, newValue) => {
        const id = get(selectedSessionIdState);

        set(sessionDataState(id), newValue);
    }
});

/*
 *  A selectorFamily that returns the data for a given variable from the selected session's data
 */
export const selectedSessionDataLengthVariableState = selectorFamily<number, Variable>({
    key: "selectedSessionDataLengthVariable",
    get: (variable: Variable) => ({get}) => {
        return get(selectedSessionDataState)[variable].length;
    }
});

/*
 *  A selectorFamily that returns the last value for a given variable from the selected session's data
 */
export const selectedSessionLastValueState = selectorFamily<number | null, Variable>({
    key: "selectedSessionLastValue",
    get: (variable: Variable) => ({get}) => {
        const data = get(selectedSessionDataState)[variable];

        if (data.length > 0) {
            return data.slice(-1)[0][1];
        } else {
            return null;
        }
    }
});

export interface VariableDisplay {
    active: boolean;
    display: "line" | "numeric";
}

export interface ActiveContainers {
    [Variable.CognitiveLoad]: VariableDisplay;
    [Variable.PerceivedDifficulty]: VariableDisplay;
    [Variable.Familiarity]: VariableDisplay;
    [Variable.InformationProcessingIndex]: VariableDisplay;
    [Variable.PhysiologicalArousal]: VariableDisplay;
    [Variable.Engagement]: VariableDisplay;
    [Variable.PhysiologicalStress]: VariableDisplay;
    [Variable.EmotionalRegulation]: VariableDisplay;
    [Variable.MotionStability]: VariableDisplay;
    [Variable.EnergySpentFatigue]: VariableDisplay;
}

/*
 *   An atomFamily that stores the active containers for each session
 */
export const sessionActiveContainersState = atomFamily<ActiveContainers, number | null>({
    key: "sessionActiveContainers",
    default: {
        [Variable.CognitiveLoad]: {active: false, display: "line"},
        [Variable.PerceivedDifficulty]: {active: false, display: "line"},
        [Variable.Familiarity]: {active: false, display: "line"},
        [Variable.InformationProcessingIndex]: {active: false, display: "line"},
        [Variable.PhysiologicalArousal]: {active: false, display: "line"},
        [Variable.Engagement]: {active: false, display: "line"},
        [Variable.PhysiologicalStress]: {active: false, display: "line"},
        [Variable.EmotionalRegulation]: {active: false, display: "line"},
        [Variable.MotionStability]: {active: false, display: "line"},
        [Variable.EnergySpentFatigue]: {active: false, display: "line"}
    }
});

/*
 *  A selector for getting and setting display object for all containers in a session
 */
export const selectedSessionActiveContainersState = selector<ActiveContainers>({
    key: "selectedSessionActiveContainers",
    get: ({get}) => {
        const id = get(selectedSessionIdState);
        return get(sessionActiveContainersState(id));
    },
    set: ({get, set}, newValue) => {
        const id = get(selectedSessionIdState);
        set(sessionActiveContainersState(id), newValue);
    }
});

/*
 *  A selectorFamily that accesses a sessions containers active state
 */

export const selectedSessionVariableContainerVisibleState = selectorFamily<VariableDisplay, Variable>({
    key: "selectedSessionVariableContainer",
    get: (variable: Variable) => ({get}) => {
        const id = get(selectedSessionIdState);
        return get(sessionActiveContainersState(id))[variable];
    },
    set: (variable: Variable) => ({get, set}, newValue) => {
        const id = get(selectedSessionIdState);
        set(sessionActiveContainersState(id), (activeContainerPrevState) => {
            return {
                ...activeContainerPrevState,
                [variable]: newValue
            };
        });
    }
});

export const snackOpenState = atom<boolean>({
    key: "snackOpen",
    default: false
});

/**
 * An atom family that stores the layouts for all current sessions
 */

export const dashboardLayoutsState = atomFamily<Layouts | undefined, number | null>({
    key: "dashboardLayouts",
    default: undefined
});

/*
 *  A selector for getting and setting dashboard layout for the currently selected session
 */
export const selectedSessionDashboardLayoutsState = selector<Layouts | undefined>({
    key: "selectedSessionDashboardLayouts",
    get: ({get}) => {
        const id = get(selectedSessionIdState);
        return get(dashboardLayoutsState(id));
    },
    set: ({get, set}, newValue) => {
        const id = get(selectedSessionIdState);
        set(dashboardLayoutsState(id), newValue);
    }
});

/**
 * An atom family that stores the column count for each session dashboard
 */

export const dashboardColumnsState = atomFamily<number | undefined, number>({
    key: "dashboardColumns",
    default: () => {
        let cols = 1;
        if (window.innerWidth >= 1200) {
            cols = 6;
        } else if (window.innerWidth >= 996) {
            cols = 4;
        } else if (window.innerWidth >= 480) {
            cols = 2;
        }
        return cols;
    }
});

/*
 *  A selector for getting and setting the number of dashboard columns for the currently selected session
 */
export const selectedSessionDashboardColumnsState = selector<number | undefined>({
    key: "selectedSessionDashboardColumns",
    get: ({get}) => {
        const id = get(selectedSessionIdState);
        if (id) {
            return get(dashboardColumnsState(id));
        }
    },
    set: ({get, set}, newValue) => {
        const id = get(selectedSessionIdState);
        if (id) {
            set(dashboardColumnsState(id), newValue);
        }
    }
});

export interface allsessionsObject {
    sessionId: number;
    studentName: string;
    data: [number, number][];
}

/*
 *  A selectorFamily that returns the data for a given variable and session id
 */
export const sessionVariableDataState = selectorFamily<[number, number][], [Variable, number]>({
    key: "sessionVariableData",
    get: (Parameters: [Variable, number]) => ({get}) => {
        return get(sessionDataState(Parameters[1]))[Parameters[0]];
    }
});

/*
 *  An atom that stores which variable is selected for the all sessions view
 */
export const selectedAllSessionVariableState = atom<Variable | null>({
    key: "selectedAllSessionVariable",
    default: null
});

/*
 *  A selector that returns array of objects that each contain student name and session data for
 *  the currently selected variable.
 */
export const allSessionsState = selector<allsessionsObject[]>({
    key: "allSessions",
    get: ({get}) => {
        const result: allsessionsObject[] = [];
        const sessions = get(sessionsState);
        const selectedVariable = get(selectedAllSessionVariableState);
        sessions.forEach((session) => {
            if (selectedVariable != null) {
                const data = get(sessionVariableDataState([selectedVariable, session._id]));
                result.push({sessionId: session._id, studentName: session.student.name, data: data});
            }
        });
        return result;
    }
});
