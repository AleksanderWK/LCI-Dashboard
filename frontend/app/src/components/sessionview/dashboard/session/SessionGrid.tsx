import {createStyles, makeStyles} from "@material-ui/core";
import React, {useMemo} from "react";
import GridLayout, {Layout, WidthProvider} from "react-grid-layout";
import {useRecoilCallback, useRecoilState, useRecoilValue} from "recoil";
import {Variable} from "../../../../constants";
import {
    selectedSessionActiveContainersState,
    selectedSessionLayoutItemState,
    selectedSessionLayoutState
} from "../../../../state/dashboard";
import packer from "../../../../utils/packer";
import {ResizeIcon} from "../../../common/Icons";
import Container from "./Container";

const useStyles = makeStyles(() =>
    createStyles({
        resizeIcon: {
            position: "absolute",
            right: "4px",
            bottom: "4px",
            width: "10px",
            height: "10px",
            "&:hover": {
                cursor: "se-resize"
            }
        }
    })
);

const StaticGridLayout = WidthProvider(GridLayout);

/**
 * The grid layout for the live session view (a specific session).
 * Renders draggable and resizable containers.
 */
export default function SessionGrid(): JSX.Element {
    const classes = useStyles();

    const selectedSessionActiveContainers = useRecoilValue(selectedSessionActiveContainersState);

    const [selectedSessionLayout, setSelectedSessionLayout] = useRecoilState(selectedSessionLayoutState);

    /**
     * Get the layout from state (position, width and height) for a given variable
     */
    const getLayoutItem = useRecoilCallback(({snapshot}) => (variable: Variable): Layout | undefined => {
        return snapshot.getLoadable(selectedSessionLayoutItemState(variable)).getValue();
    });

    /**
     * Renders memoized containers for each selected variable
     * The containers are rerendered only if the selected containers is changed (one is added or removed)
     */
    const containers = useMemo(() => {
        const layout: Layout[] = [];
        return selectedSessionActiveContainers.map((variable) => {
            const existingItem = getLayoutItem(variable);
            const item = existingItem
                ? existingItem
                : {i: variable, w: 4, h: 3, minW: 2, minH: 3, ...packer(4, 3, layout, 12)};
            layout.push(item);
            return (
                <div key={variable} data-grid={item}>
                    <Container variable={variable} />
                </div>
            );
        });
    }, [selectedSessionActiveContainers]);

    return (
        <StaticGridLayout
            cols={12}
            rowHeight={60}
            containerPadding={[0, 0]}
            margin={[10, 10]}
            draggableCancel=".noDrag"
            isBounded
            resizeHandle={<ResizeIcon className={`${classes.resizeIcon} noDrag`} color="action" />}
            layout={selectedSessionLayout}
            onLayoutChange={(layout) => {
                // Check if the previous layout is different to the new one before updating state
                // This prevents unnecessary rerenders
                if (JSON.stringify(layout) !== JSON.stringify(selectedSessionLayout)) {
                    setSelectedSessionLayout(layout);
                }
            }}
        >
            {containers}
        </StaticGridLayout>
    );
}
