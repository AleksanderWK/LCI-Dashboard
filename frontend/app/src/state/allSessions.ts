/*
 *  State used in the all sessions view
 */

import {atom, selector} from "recoil";
import {Variable} from "../constants";
import {EducationalSpecificEmotions, sessionsState, sessionVariableDataState} from "./session";

export interface allsessionsObject {
    sessionId: number;
    studentName: string;
    data: [number, number | EducationalSpecificEmotions][];
}

/*
 *  An atom that stores which variable is selected for the all sessions view
 */
export const selectedAllSessionVariableState = atom<Variable | null>({
    key: "selectedAllSessionVariable",
    default: null
});

/*
 *  A selector that returns array of objects that each contain student name and session data for
 *  the currently selected variable.
 */
export const allSessionsState = selector<allsessionsObject[]>({
    key: "allSessions",
    get: ({get}) => {
        const result: allsessionsObject[] = [];
        const sessions = get(sessionsState);
        const selectedVariable = get(selectedAllSessionVariableState);
        sessions.forEach((session) => {
            if (selectedVariable != null) {
                const data = get(sessionVariableDataState([selectedVariable, session._id]));
                result.push({sessionId: session._id, studentName: session.student.name, data: data});
            }
        });
        return result;
    }
});
