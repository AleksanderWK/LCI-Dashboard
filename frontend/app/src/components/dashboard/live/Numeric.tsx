import React from "react";
import {createStyles, makeStyles, Theme, Typography} from "@material-ui/core";
import {useRecoilValue} from "recoil";
import {Variable} from "../../../constants";
import {selectedSessionLastValueState} from "../../../state/session";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        numericWrapper: {
            width: "100%",
            padding: "20px 0"
        },
        numeric: {
            fontSize: 50,
            color: theme.palette.text.default,
            textAlign: "center"
        },
        noData: {
            color: theme.palette.text.default,
            textAlign: "center"
        }
    })
);

interface Props {
    variable: Variable;
}

function Numeric(props: Props): JSX.Element {
    const classes = useStyles();

    // Get the latest variable data point for the selected user
    const dataPoint = useRecoilValue(selectedSessionLastValueState(props.variable));

    return (
        <div className={classes.numericWrapper}>
            {dataPoint ? (
                <Typography className={classes.numeric}>{dataPoint}</Typography>
            ) : (
                <Typography className={classes.noData}>No data received</Typography>
            )}
        </div>
    );
}

export default Numeric;
