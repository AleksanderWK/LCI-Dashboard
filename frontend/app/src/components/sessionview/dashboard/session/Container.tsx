import React, {useRef, useState} from "react";
import {MMDVariables, Variable} from "../../../../constants";
import Menu from "../Menu";
import LineChart from "../LineChart";
import Numeric from "../Numeric";
import ContainerCard from "../../../common/ContainerCard";
import Tooltip from "../../../common/Tooltip";
import {makeStyles, Theme, createStyles, IconButton, Typography} from "@material-ui/core";
import InfoOutlinedIcon from "@material-ui/icons/InfoOutlined";
import MoreVertIcon from "@material-ui/icons/MoreVert";
import CalculatingIndicator from "../CalculatingIndicator";
import {useRecoilState, useRecoilValue, useSetRecoilState} from "recoil";
import {selectedSessionDataLengthVariableState, selectedSessionIdState} from "../../../../state/session";
import XRangeChart from "../XRangeChart";
import {viewState, selectedSessionActiveContainersState, selectedSessionLayoutState} from "../../../../state/dashboard";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        header: {
            display: "grid",
            gridTemplateColumns: "1fr 56px",
            gap: theme.spacing(2)
        },
        menu: {
            display: "grid",
            gridTemplateColumns: "24px 24px",
            gap: theme.spacing(1)
        },
        iconButton: {
            padding: 0
        },
        infoIcon: {
            cursor: "default"
        }
    })
);

interface Props {
    variable: Variable;
}

export default function Container(props: Props): JSX.Element {
    const classes = useStyles();

    const dataLength = useRecoilValue(selectedSessionDataLengthVariableState(props.variable));

    const selectedSessionId = useRecoilValue(selectedSessionIdState);
    const [view, setView] = useRecoilState(viewState([selectedSessionId, props.variable]));
    const setActiveContainers = useSetRecoilState(selectedSessionActiveContainersState);
    const setSelectedSessionLayout = useSetRecoilState(selectedSessionLayoutState);

    const [menuOpen, setMenuOpen] = useState<boolean>(false);

    const menuAnchorElement = useRef<HTMLDivElement | null>(null);

    function removeContainer(): void {
        setActiveContainers((prevState) => {
            const newState = [...prevState];
            return newState.filter((item) => item != props.variable);
        });

        // Remove from layout
        setSelectedSessionLayout((prevValue) => {
            const layout = [...prevValue];
            return layout.filter((item) => item.i != props.variable);
        });
    }

    return (
        <ContainerCard>
            <>
                <div className={classes.header}>
                    <Typography variant="h2" noWrap={true}>
                        {MMDVariables[props.variable].name}
                    </Typography>

                    <div className={`${classes.menu} noDrag`} ref={menuAnchorElement}>
                        <Tooltip variable={props.variable}>
                            <IconButton
                                aria-label="info"
                                disableFocusRipple={true}
                                disableRipple={true}
                                disableTouchRipple={true}
                                className={`${classes.iconButton} ${classes.infoIcon}`}
                            >
                                <InfoOutlinedIcon color="action" />
                            </IconButton>
                        </Tooltip>
                        <IconButton
                            aria-label="settings"
                            className={classes.iconButton}
                            onClick={() => setMenuOpen(true)}
                        >
                            <MoreVertIcon color="action" />
                        </IconButton>
                    </div>

                    <Menu
                        open={menuOpen}
                        anchorEl={menuAnchorElement.current}
                        isDetailedView={view === "chart"}
                        onShowMore={() => setView("chart")}
                        onShowLess={() => setView("numeric")}
                        onRemoveView={() => removeContainer()}
                        onMenuClose={() => setMenuOpen(false)}
                    />
                </div>
                {MMDVariables[props.variable].calculationTime && dataLength === 0 ? (
                    <CalculatingIndicator variable={props.variable} />
                ) : view === "chart" && props.variable != Variable.EducationalSpecificEmotions ? (
                    <LineChart variable={props.variable} />
                ) : view === "chart" ? (
                    <XRangeChart variable={props.variable} />
                ) : (
                    <Numeric variable={props.variable} />
                )}
            </>
        </ContainerCard>
    );
}
