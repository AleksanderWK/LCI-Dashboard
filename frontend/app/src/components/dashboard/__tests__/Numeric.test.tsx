import React from "react";
import ReactDOM from "react-dom";
import {render, screen} from "@testing-library/react";
import Numeric from "../Numeric";
import {RecoilRoot} from "recoil";
import {Variable} from "../../../constants";
import {sessionDataState} from "../../../state/session";

it("matches snapshot", () => {
    const {baseElement} = render(
        <RecoilRoot>
            <Numeric variable={Variable.CognitiveLoad} />
        </RecoilRoot>
    );
    expect(baseElement).toMatchSnapshot();
});

it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
        <RecoilRoot>
            <Numeric variable={Variable.CognitiveLoad} />
        </RecoilRoot>,
        div
    );
});

it("shows correct value when no data received", () => {
    render(
        <RecoilRoot>
            <Numeric variable={Variable.CognitiveLoad} />
        </RecoilRoot>
    );

    expect(screen.getByText(/no data received/i)).toBeInTheDocument();
});

it("shows correct value", () => {
    render(
        <RecoilRoot
            initializeState={(snap) =>
                snap.set(sessionDataState(null), {
                    [Variable.CognitiveLoad]: [[new Date().getTime(), 50]],
                    [Variable.PerceivedDifficulty]: [],
                    [Variable.Familiarity]: [],
                    [Variable.InformationProcessingIndex]: [],
                    [Variable.PhysiologicalArousal]: [],
                    [Variable.Engagement]: [],
                    [Variable.PhysiologicalStress]: [],
                    [Variable.EmotionalRegulation]: [],
                    [Variable.MotionStability]: [],
                    [Variable.EnergySpentFatigue]: []
                })
            }
        >
            <Numeric variable={Variable.CognitiveLoad} />
        </RecoilRoot>
    );

    expect(screen.getByText(/50/i)).toBeInTheDocument();
});
