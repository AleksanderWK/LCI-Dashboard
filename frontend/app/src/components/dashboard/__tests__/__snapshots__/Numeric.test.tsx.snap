// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<body>
  <div>
    <div
      class="makeStyles-numericWrapper-1"
    >
      <p
        class="MuiTypography-root makeStyles-noData-3 MuiTypography-body1"
      >
        No data received
      </p>
    </div>
  </div>
</body>
`;
